rir/src/compiler/opt/elide_env_spec.cpp

```diff
@@ -11,62 +11,72 @@ namespace rir {
 namespace pir {
 
 void ElideEnvSpec::apply(RirCompiler&, Closure* function) const {
+    ProfiledValues& profile = function->runtimeFeedback;
 
     // Elide environments of binary operators in which both operators are
     // primitive values
     std::unordered_map<BB*, Checkpoint*> checkpoints;
 
-    auto insertExpectAndAdvance = [&](BB* src, Value* operand,
-                                      BB::Instrs::iterator position) {
-        Instruction* condition = new IsObject(operand);
-
-        position = src->insert(position, condition);
-        position++;
+    auto insertExpect = [&](BB* src, Value* condition, Value* operand,
+                            Checkpoint* cp, BB::Instrs::iterator position) {
         assert(checkpoints[src]);
-        position =
-            src->insert(position, new assumeNot(condition, checkpoints[src]));
+        position = src->insert(position, (new Assume(condition, cp))->Not());
         position++;
         return position;
     };
+    auto insertExpectNotObj = [&](BB* src, Value* operand, Checkpoint* cp,
+                                  BB::Instrs::iterator position) {
+        auto condition = new IsObject(operand);
+        position = src->insert(position, condition);
+        position++;
+        return insertExpect(src, condition, operand, cp, position);
+    };
 
     Visitor::run(function->entry, [&](BB* bb) {
         if (bb->isEmpty())
             return;
-
         if (auto cp = Checkpoint::Cast(bb->last()))
             checkpoints.emplace(bb->trueBranch(), cp);
+    });
 
+    Visitor::run(function->entry, [&](BB* bb) {
         if (!checkpoints.count(bb))
             return;
+        auto cp = checkpoints.at(bb);
 
         auto ip = bb->begin();
         while (ip != bb->end()) {
             Instruction* i = *ip;
+            auto next = ip + 1;
```

Comment:
why?

Reply:
why do you think so?

Reply:
But `ip` has a different value inside the if... Granted here we could just use `ip++` at the end. But I had so many subtle bugs with iterators, that I started using the same robust pattern everywhere:

```
while (ip != end) {
  auto next = ip+1;
  ...
  ip = next;
}
```

Now you are suggesting:

```
while (ip != end) {
  ...
  ip++;
}
```

Let me show you how this breaks:

At some point I decide to add a new case to this loop. Namely:
```
...
  ip = bb->remove(ip);
  ip++;
```

so now, what happens if I remove the last element?
the `ip` points to the previously second-last after removal. But since I don't want to process this instruction twice i add the `ip++`, so no `ip` points to the end.

Then at the end of the loop the `ip++` actually increments over the end, the loop condition is true and we get a segv (maybe).

Reply:
The same think does not happen with the above pattern. I would add the case like this:
```
  ip = bb->remove(ip);
  next = ip+1
```

and since we are not incrementing next anymore, this will do the right thing

Reply:
I do not see why in:
```
...
  ip = bb->remove(ip);
  ip++; 
```
you do `ip++` if you know that the end of the loop will do that.

Anyway, ok if we have a new pattern for iterations. I didn't know about it at the moment of the comment.